Upgrades

eric-winkler opened the pull request:
Addressing https://github.com/LightBlueProject/LightBlue/issues/144

eric-winkler committed:
Target 4.8

eric-winkler committed:
New csproj format - LightBlue.csproj

eric-winkler committed:
New csproj format - LightBlue.autofac

eric-winkler committed:
New csproj format - LighBlue.Host.Stub

eric-winkler committed:
New csproj format - LightBlue.Host

eric-winkler committed:
New csproj format - LightBlue.MultiHost

eric-winkler committed:
New csproj format - LightBlue.Testability

eric-winkler committed:
New csproj format - LightBlue.Tests

eric-winkler committed:
Upgrade Xunit 1.9.2 -> 2.4.2

eric-winkler committed:
New csproj format - LightBlue.WebHost

eric-winkler committed:
New csproj format - LightBlue.WebService

eric-winkler committed:
New csproj format - LightBlue.Windsor

eric-winkler committed:
New csproj format - LightBlue.WorkerRoleDependency

eric-winkler committed:
New csproj format - LightBlue.WorkerService

eric-winkler committed:
New csproj format - TestWorkerRole

eric-winkler committed:
Remove LightBlue.Windsor

eric-winkler committed:
New csproj format - TestWebRole

eric-winkler committed:
Version is in csproj now

eric-winkler committed:
Don't track the VisualStudio settings

eric-winkler committed:
Modernise use of nuget

eric-winkler committed:
Upgrade build image

eric-winkler committed:
Ensure compatability with csproj version number format

eric-winkler committed:
The PackageLicenceUrl element is deprecated

eric-winkler committed:
Extend the years in the copyright message

eric-winkler committed:
(WIP) Netstandard LightBlue

eric-winkler committed:
Upgrade Azure storage library to v12

eric-winkler committed:
Replace Newtonsoft.Json with System.Text.Json

eric-winkler committed:
Refresh tests for v12 storage upgrade

eric-winkler committed:
Upgrade Autofac 3.5.2->6.5.0

eric-winkler committed:
Bumping to v4 due to API change

eric-winkler committed:
Add support for `dotnet pack`

Why are the tools missing frmo LightBlue.MultiHost package?

seanswallingcc commented on LightBlue/Standalone/StandaloneList.cs:
```diff
@@ -1,37 +1,26 @@
-            var resultSegment = blobListing == BlobListing.Flat
-                ? FindFilesFlattened(containerDirectory, searchDirectory, prefix, maxResults, numberToSkip)
-                : FindFilesHierarchical(containerDirectory, searchDirectory, prefix, maxResults, numberToSkip);
+            var resultSegment = FindFilesFlattened(containerDirectory, prefix, maxResults);
```
I assume 'Listing Snapshots' and 'Hierarchical Mode' are not used and are now unsupported?

seanswallingcc replied:
As well as 'Continuation' and 'Skipping'?

eric-winkler replied:
Hierachical listing was not used in our code bases so I've deprecated support for the feature.

Snapshot listing should still be possible by passing `BlobStates.Snapshots`

The underlying API has adopted AsyncEnumerable to support continuation. Our code bases were always performing full enumerations so I took the lazy approach and just implemented full enumeration.

seanswallingcc commented on LightBlue/Standalone/StandaloneMetadataStore.cs:
```diff
@@ -1,51 +1,31 @@
-
-using Newtonsoft.Json;
-using Newtonsoft.Json.Linq;
+using System.Text.Json;
```
Just double checking against the csproj and package.config (now deleted), LightBlue.csproj declares this binary as standard2.0, but standard2.0 doesn't include System.Text.Json since it's only bundled with .net3.0+.
net48 consumers woudl have to pull that package in, but I don't see it listed as a dependency of lightblue.csproj.
Did I miss something?

eric-winkler replied:
It must be coming in as a transitive dependency somewhere. Is it best practice to explicitly take the dependency?

seanswallingcc replied:
I think if you depend directly on it, it's best to take that dependency explicitly.  The knock on issue is consumers need a way for nuget to pull in stj so LightBlue can run.  This used to be done with the nuspec file, but I think it can be done with csproj now.  The trick I think is to not let consumers propagate the transitive dependency they get from LB so they can pick what they want, while still allowing LB to work.

seanswallingcc reviewed (commented):
Just need some clarification around features and stj.

seanswallingcc committed:
Hide away Azure Storage concerns

eric-winkler committed:
Remove unneeded Newtonsoft.Json dependency

eric-winkler committed:
Merge pull request #1 from seanswallingcc/HideAzureStorage

Hide away Azure Storage concerns

seanswallingcc committed:
Allow multiple readers of metadatastore

eric-winkler committed:
Merge pull request #2 from seanswallingcc/FixBlockingReads

Allow multiple readers of metadatastore

seanswallingcc committed:
Hide away Azure Queue concerns

seanswallingcc committed:
swap BlobCopyStatus for LightBlueBlobCopyStatus

eric-winkler committed:
Merge pull request #4 from seanswallingcc/SwapBlobCopyStatus

Swap blob copy status

eric-winkler committed:
Disabling test concurrency

seanswallingcc committed:
Upgrade Autofac 6.5.2 -> 8.0.0

eric-winkler committed:
Merge pull request #5 from seanswallingcc/BumpAutofac

Upgrade Autofac 6.5.2 -> 8.0.0

seanswallingcc committed:
fix: performance

- make metadata reading async
- fetching data from FileStream can be pretty slow at scale, even when
fetching from it's internal buffer.  By keeping the StreamReader's
buffer the same size as the FileStream, it should only need one fetch
from the FileStream.

seanswallingcc commented on LightBlue/Hosted/HostedAzureBlockBlob.cs:
```diff
@@ -1,30 +1,47 @@
         public HostedAzureBlockBlob(Uri blobUri)
         {
-            _cloudBlockBlob = new CloudBlockBlob(blobUri);
+            _cloudBlockBlob = new BlockBlobClient(blobUri);
```
Setting of _properties is optional in the ctors here, but Public Properties `CopyState` and `Metadata` will hit a null reference when being used in that scenario.  Mercury for example when saving Blobs will hit this issue.

eric-winkler committed:
Fix the ability to add metadata

Previous changes had broken the access paradigm, which is that metadata can be added to a new HostedAzureBlockBlob via `blob.Metadata.Add(...)`
Additionally, metadata & content type weren't being correctly included in blob upload requests.

Commit:
Merge pull request #6 from seanswallingcc/FixPerformanceIssue

fix: performance

eric-winkler committed:
Include the appconfig in the tools folder

It contains useful binding redirects.

seanswallingcc commented on LightBlue/Standalone/StandaloneAzureBlockBlob.cs:
```diff
@@ -321,15 +327,23 @@ public Task UploadFromByteArrayAsync(byte[] buffer)
-            }
+                throw new ArgumentNullException(nameof(buffer));
+
+            if (index < 0 || index >= buffer.Length)
```
We've found this throws when calling `UploadFromByteArrayAsync` with an empty `byte[]`

Mercury appears to trigger this in at least one scenario, but currently unsure if it's unintentional, or expected and whether or not we should support it.  Still investigating.

seanswallingcc replied:
I can confirm Mercury does this somewhat intentionally.  Or at least expects to be able to upload empty blobs.

eric-winkler replied:
I've fixed the overzealous argument validation and added a missing test.

eric-winkler committed:
Support uploading empty blobs

seanswallingcc committed:
fix: push null results to consumer like before

Commit:
Merge pull request #7 from seanswallingcc/push-null-fix

fix: push null results to consumer like before

eric-winkler committed:
Ensure blob properties are populated on download

seanswallingcc committed:
fix: blob properties

Ensure properties are are populated on download for standalone
implementation

Commit:
Merge pull request #8 from seanswallingcc/Upgrades

fix: blob properties

eric-winkler committed:
Internal asyncs should ConfigureAwait(false)

eric-winkler committed:
Ensure the blob properties are populated

This may have a performance consequence as OpenRead has some read buffering

eric-winkler committed:
Allow Content Type to be set in one call

eric-winkler committed:
Ensure blob properties are populated on download

eric-winkler committed:
Upgrading dependencies

eric-winkler committed:
Align StandaloneAzureQueue behavior with Hosted

Maintains original AsString contract on the queue message

eric-winkler committed:
Upgrade to MahApps.Metro 2.4.10

eric-winkler committed:
Swap deprecated NDesk.Options for Spectre.Console
